src/physics/initial_conditions/initial_condition_base.h

```diff
@@ -0,0 +1,53 @@
+#ifndef __INITIAL_CONDITION_BASE_H__
+#define __INITIAL_CONDITION_BASE_H__
+
+#include <deal.II/lac/vector.h>
+#include <deal.II/base/function.h>
+#include "parameters/all_parameters.h"
+#include "parameters/parameters.h"
+#include "dg/dg.h"
+#include "initial_condition_function.h"
+
+namespace PHiLiP {
+
+///Initial Condition Base class.
+
+template <int dim, int nstate, typename real>
+class InitialConditionBase
```

Comment:
What does "Base" in your InitialConditionBase class mean? Are you planning on deriving more classes from this?

How is it different from the dealii::VectorTools::project and dealii::VectorTools::interpolate function?

Reply:
But your various initial conditions depend on the different InitialConditionFunction. InitialConditionBase doesn't seem to have anything else that would be derived from it. Right?

I see now that you're providing the reference cell quadrature point to the InitialConditionFunction instead of the physical location of that quadrature node, hence why you're doing a projection in the reference space rather than the physical and thus not using dealii::project. But.... this now means that your InitialConditionFunction is the same function for every single cell (which is then curved by the grid metrics when computing residuals), which is extremely counterintuitive. I think it's a fair assumption that people would provide a InitialConditionFunction that represents the physical solution distribution throughout the entire domain, and that your projection is a special case specific for you paper. Otherwise, your initial conditions basically vary with your initial grid...

e.g. 
Your projection:
Providing initial_conditions = sin(2pi*x), on a physical domain between 0...1, with 2 cells, you'd end up with 2 full periods. With 3 cells, 3 full periods. 

The usual projection (as defined with dealii::VectorTools::project())
Providing initial_conditions = sin(2pi*x), on a physical domain between 0...1, with 2 cells, you'd end up with 1 full periods. With 3 cells, 1 full period.

DG does integration in the reference space, but the solution distribution is still valid in the physical space.

Reply:
Ok you are correct about the nothing being deirved from it.
No, I am not passing the reference quadrature point. fe_values.quadrature_point(iquad) is the physical location of the quadrature point.

Reply:
OK, then why do you use a reference projection rather than a physical projection? This means that if you have two different curved grids for the same geometry, your initial conditions will be different.

Reply:
We use a reference projection because we assume the solution u is spanned by the reference basis functions. ie/ interp(u) to the quadrature nodes in the reference space should recover the initial condition at the physical quadrature nodes. The physical projection with J built in it would only be used if we span J*u by the basis (note this is used by J. Chan for the weight-adjusted work and Z.J. Wang in CPR). 
Yes 2 different curved grids should give different solution modal coefficients for the same overall geometry.

Reply:
Right. So, if you assume that the solution is spanned by the basis function, then there is no point in doing a projection. The projection is equivalent to an interpolation. J*u would never be spanned by the basis since J is rational. 

If you think about initial conditions, they are defined by your non-discretized equations. That is, f(x,t=0) = sin(2pi*(x-t)) would be the physical initial conditions of your problem. By defining your reference projection, your are effectively changing the unsteady problem to be solved depending on your grid curvatures.

Basically, if I want to do a vortex advection problem, this reference projection would introduce an error in the initial conditions. Not sure in what scenario you are using your current projection.

Reply:
I disagree entirely. Let's write this out and assume the initial condition is u(x,0)=e^(x) (at the volume nodes).
I will use " |_{} " to state where it is evaluated
(a) My reference projection: \hat{u}^T = projection( e^{x}|_{at physcial flux nodes} ) , such that interp(\hat{u}^T)_{flux nodes} = e^x. 
(b) dealii::interpolate: \hat{u}^T = e^x |_{solution nodes}. Note that interp(e^x |_{solution nodes})_{flux nodes} != (a) because of the nonlinearity in e^x. 
(c) dealii::project: \hat{u}^T = (\chi^T *W*J*\chi)^{-1} * \chi^T * W * J * e^{x}|_{flux nodes} ,  note that interp(\hat{u}^T)_{flux nodes} != e^x since the Jacobian is mixed in there.

For the linear case all 3 above are equivalent.

Reply:
Yup, I'm aware of (a), (b), and (c). The main thing is that neither (a), nor (b) minimizes the error in the initial condition. The fact that you recover e^x when you reinterpolate the resulting solution in case (a) is a red flag that you haven't actually minimized the the solution. You might as well just query e^x at your flux nodes, and use a Lagrange polynomial. 

I understand that in case (c) you don't recover the solution at the queried points, but that's a good thing. You want to minimize the functional L2 norm, not minimize the error at those points. There is a reason why dealii::interpolate and dealii::project() are defined, but a reference projection doesn't.

Another guess of why you are doing this is because you are assuming that it's a "discrete projection" and you're basically underintegrating your projection to recover exactly the initial condition values at your flux nodes. This would explain why you're somehow specifying nodes, when the exact mathematical projection e^x onto a polynomial basis does not depend on nodes. Projecting f(x) = e^x, onto the basis (1,x), on the domain [0,1], does not depend on nodes. As you integrate more and more exactly, you will converge to a single set of coefficient {a,b} such that the error function (e^x - a*1 + b*x) is orthogonal to your basis (1,x).

Have you tried using dealii::project() in your test case?

Reply:
Ok, after carefully going through it again, I agree that you are correct.
What happened before with the reference projection is that since the basis was not overintegrated, then the basis function inverse is exact and the determinants of the Jacobians cancel off. Which is the result from your "discrete projection" explanation above. 
I will change this to using dealii::project to handle the overintegrated case.
Thank you !

Reply:
Yeah, I'm pretty sure you can recover the same "discrete projection" or reference projection by calling dealii::interpolate if it matters for your paper. I just think it would be a third type that is not typical and not exactly called a simple "projection".